src/main/java/com/example/planit/engine/CalendarEngine.java

```diff
@@ -1402,22 +1402,94 @@ public static boolean hasAuthorizeScopesStillValid(String access_token) throws I
         return false; // scope was not found or response wasn't successful
     }
 
-    public DTOstudyPlanResponseToController getUserLatestStudyPlan(String sub) {
+    public DTOstudyPlanAndUpcomingSessionResponseToController getUserLatestStudyPlan(String sub) {
 
         try {
 
             Optional<User> maybeUser = userRepo.findUserBySubjectID(sub);
 
             if (maybeUser.isEmpty()) {
-                return new DTOstudyPlanResponseToController(false, ERROR_USER_NOT_FOUND, HttpStatus.BAD_REQUEST);
+                return new DTOstudyPlanAndUpcomingSessionResponseToController(false, ERROR_USER_NOT_FOUND, HttpStatus.BAD_REQUEST);
             }
-            return new DTOstudyPlanResponseToController(true, NO_PROBLEM, HttpStatus.OK, maybeUser.get().getLatestStudyPlan());
+            return new DTOstudyPlanAndUpcomingSessionResponseToController(true, NO_PROBLEM, HttpStatus.OK, maybeUser.get().getLatestStudyPlan());
         } catch (Exception e) {
             logger.error(buildExceptionMessage(e));
-            return new DTOstudyPlanResponseToController(false, ERROR_DEFAULT, HttpStatus.INTERNAL_SERVER_ERROR);
+            return new DTOstudyPlanAndUpcomingSessionResponseToController(false, ERROR_DEFAULT, HttpStatus.INTERNAL_SERVER_ERROR);
         }
     }
 
+
+    public DTOstudyPlanAndUpcomingSessionResponseToController getUserUpcomingSession(String sub) {
+        try {
+
+            Optional<User> maybeUser = userRepo.findUserBySubjectID(sub);
+
+            if (maybeUser.isEmpty()) {
+                return new DTOstudyPlanAndUpcomingSessionResponseToController(false, ERROR_USER_NOT_FOUND, HttpStatus.BAD_REQUEST);
+            }
+            User user = maybeUser.get();
+            // refresh access_token before making api call
+            validateAccessTokenExpireTime(user);
+
+            // check if access_token still have scope for Google calendar
+            if (!hasAuthorizeScopesStillValid(user.getAuth().getAccessToken())) {
+                return new DTOstudyPlanAndUpcomingSessionResponseToController(false, Constants.ERROR_NO_VALID_ACCESS_TOKEN, HttpStatus.UNAUTHORIZED);
+            }
+
+            return new DTOstudyPlanAndUpcomingSessionResponseToController(true, NO_PROBLEM, HttpStatus.OK,upcomingSessionForUser(user));
+        } catch (Exception e) {
```

Comment:
i am not sure i like this approach, since you return OK, and later call a function that can fail, please refactor and make sure maybe with try catch block or something to make sure you are holding valid session and return it directly

like this :
try {
session = getSession(user)
return new DTO(session)

} catch {
return error
}

Reply:
also add "get" prefix to function call.